### ISSUE TITLE
ZookeeperLockRegistry does not honor timeout semantics on tryLock

### ISSUE THREAD

**Issue by cah-oster** [spring-projects/spring-integration#1863]
I was trying to write a test that checked the behavior of ZookeeperLockRegistry when the zookeeper cluster is unreachable.  I wrote a test that extended the pattern in DefaultLockRegistryTests to ensure the lock is not able to be locked when the testingserver is shutdown.  Here's a snippet of the code:

```
testingServer.stop();
Lock lock2 = registry.obtain("bar");
assertFalse("Should not have been able to lock with zookeeper server stopped!", lock2.tryLock(2, TimeUnit.SECONDS));

```

The call to lock2.tryLock(2, TimeUnit.SECONDS)) hangs forever until I kill the test.  This is the call in curator framework that seems to cause the problem:
https://github.com/apache/curator/blob/apache-curator-2.10.0/curator-recipes/src/main/java/org/apache/curator/framework/recipes/locks/StandardLockInternalsDriver.java#L54

These seems to occur regardless of the CuratorFramework client configuration that is passed into the registry.  I do see the background reconnect process timingout and eventually giving up (properly according to my settings), but the call linked to above seems to run in the foreground and not honor the retry/timeout settings.

**Comment by cah-oster** [comment]
While it didn't honor the semantics, I was able to get it to not retry forever by setting maxRetries on the retrypolicy on the CuratorFramework client configuration to 0.

**Comment by cah-oster** [comment]
There appears to be a timing issue relating to the above, in that sometimes its gets in an infinite recursion loop at overflows the stack:

```
at org.apache.curator.framework.imps.CuratorFrameworkImpl.checkBackgroundRetry(CuratorFrameworkImpl.java:716) [curator-framework-2.10.0.jar:na]
	at org.apache.curator.framework.imps.CuratorFrameworkImpl.performBackgroundOperation(CuratorFrameworkImpl.java:857) [curator-framework-2.10.0.jar:na]
	at org.apache.curator.framework.imps.CuratorFrameworkImpl.processBackgroundOperation(CuratorFrameworkImpl.java:507) [curator-framework-2.10.0.jar:na]
	at org.apache.curator.framework.imps.FindAndDeleteProtectedNodeInBackground$1.retriesExhausted(FindAndDeleteProtectedNodeInBackground.java:56) [curator-framework-2.10.0.jar:na]
	at org.apache.curator.framework.imps.CuratorFrameworkImpl.checkBackgroundRetry(CuratorFrameworkImpl.java:716) [curator-framework-2.10.0.jar:na]
```

**Comment by artembilan** [comment]
Sorry, what would you like to hear from us about Spring Integration here?
According to your analysis, it looks like a problem is exactly in the Curator...
 
Thanks

**Comment by cah-oster** [comment]
@artembilan I probably dug into implementation details a bit more than necessary to report it as a bug, but from a spring-integration zookeeperlockregistry user, the tryLock semantics seem broken.  I'm not familiar enough with the CuratorFramework to understand whether there are alternative calls that can/should be made by the ZkLock Lock implementation, but I'd still consider it a bug with spring-integration.  Perhaps the only resolution is "won't fix" or blocked by a bug in currator, but I can't make that assertion.

**Comment by artembilan** [comment]
OK. Share some simple test-case to play from our side, please.

**Comment by cah-oster** [comment]
@artembilan Ok, will do.  I would have already but some of the necessary test classes are not accessible outside of the project (ZookeeperTestSupport) and a test jar doesn't seem to be provided.  I'll put together a modification to ZkLockRegistryTests.java when I get a chance.

**Comment by cah-oster** [comment]
@artembilan  Here's a simple additional test I patched from master that demonstrates the behavior:
[patch.txt](https://github.com/spring-projects/spring-integration/files/404446/patch.txt)

**Comment by artembilan** [comment]
By the way, would you mind raising similar issue against Curator for gathering more info and feedback?
https://issues.apache.org/jira/browse/CURATOR

Thanks

**Comment by artembilan** [comment]
@cah-oster , well, even if that isn't our problem, using such an API is our responsibility.
IMO we should provide some workaround on the matter.
Please, see https://jira.spring.io/browse/INT-4087 and let's continue discussion there.

**Comment by artembilan** [comment]
@cah-oster , please, find the time to have a look to the provided solution #1868.
I decided to wait for the connection in the separate Thread, because keeping locked Thread isn't so simple task.
There is a tiny case when we still can lose a connection just after our `pathFor()` success result and before `mutex.acquire()`. 
But anyway we have similar race condition in the `LockInternals.attemptLock()`:
````java
ourPath = driver.createsTheLock(client, path, localLockNodeBytes);
hasTheLock = internalLockLoop(startMillis, millisToWait, ourPath);
````

**Comment by cah-oster** [comment]
@artembilan I think the approach makes sense.  At least in my case I'm only sparsely acquiring locks and just want to ensure that if Zookeeper is down when I try to do so, it only holds the thread as long as I request it to.  In a more highly contentious locking environment , I would still have concerns about the behavior during a zookeeper outage (where the second race condition is more likely), but this is certainly better than before.  Given that now the registry needs to be destroyed, I might suggest you add something to the config package that handles that, similar to the CuratorFrameworkFactoryBean which handles shutting down the CuratorClient appropriately.

**Comment by artembilan** [comment]
@cah-oster , thank you for feedback!

If you have some concerns about that change, feel free to comment on PR directly.
Regarding `destroy()`. Yes, we definitely assume that `ZookeeperLockRegistry`  is registered as a bean in the application context, where the last one will handle `DisposableBean` properly on application shutdown.
I don't see reason to introduce `FactoryBean` wrapper, since `ZookeeperLockRegistry` looks pretty straightforward.
Although I guess @garyrussell will suggest to add `setMutexTaskExecutor()` to let end-user to inject managed by the application context `TaskExecutor`. :smile: 
I won't mind, but would like to see if the solution is sufficient first of all.

I spent a half day yesterday to try to figure out Thread Pool with mapped Threads to let to re-lock and release properly. But in the end I decided that I'm going to reinvent `ThreadPoolExecutor` with all its magic `Worker` management logic. :cry:

**Comment by cah-oster** [comment]
@artembilan You're comments on destroy() make sense and works for me.  I don't have any specific concerns beyond what you already called out (the remaining chance at a race condition), but I think that's significantly less possible of a scenario than the ZK being down before the call is made, so I'd view this as a big improvement.  As long as the test I added passes, I'd say the solution is sufficient.  FYI, the travis build on the PR failed due to a checkstyle rule.  As long as that's resolved and the tests pass, I'd say I'm happy.

**Comment by artembilan** [comment]
Thank you for your help @cah-oster !

The only problem with ZK that there is no any hooks to check the real connection state.
We can know that there is a problem only when dispatch the next command to ZK.
And if we do that in our main Thread we wait for all the re-connection and recovery "hell". :cry: 

Looking forward for more valuable feedback from you or anyone else in the community.

@garyrussell , now it's your turn to take a look into PR :smile:

**Comment by garyrussell** [comment]
LGTM but, yes, we need an alternate ctor for people running on WL/WS etc - so they can inject a WMTE.
